components/encryption/src/manager/mod.rs

```diff
@@ -98,75 +116,91 @@ impl Dicts {
         }
     }
 
-    fn save_key_dict(&mut self, master_key: &dyn Backend) -> Result<()> {
+    fn save_key_dict(&self, master_key: &dyn Backend) -> Result<()> {
+        let _lock = self.key_lock.lock().unwrap();
```

Comment:
move this lock to `maybe_rotate_data_key`, and double check if current_key_id is the same, before the following lines:
```
        let duration = now.duration_since(UNIX_EPOCH).unwrap();
        let creation_time = duration.as_secs();	        let creation_time = duration.as_secs();
```